core/src/main/java/com/linecorp/armeria/internal/server/DefaultServiceRequestContext.java

```diff
@@ -255,6 +257,18 @@ public ServiceConfig config() {
         return cfg;
     }
 
+    @Nullable
+    @Override
+    public <T extends HttpService> T findService(Class<? extends T> serviceClass) {
+        requireNonNull(serviceClass, "serviceClass");
+        final HttpService service = config().service();
+        if (service instanceof InitialDispatcherService) {
+            return ((InitialDispatcherService) service).findService(this, serviceClass);
```

Comment:
I discovered a case that a decorator isn't found by this logic:
https://github.com/line/armeria/blob/a54f418ced3bf3423efebf7766f24387ca84ee9d/core/src/main/java/com/linecorp/armeria/server/ServiceConfig.java#L159
Perhaps we can get the `handlesCorsPreflight` on demand but I'm not sure if it's a good approach. Do you have any idea?

Reply:
I think the only way is to check if `CorsService` exists when a request is received.
Let me take a look closer.

Reply:
I figured out CorsService` will work well with no problem even though it may be unintentional. https://github.com/line/armeria/blob/01f2c6f6ee49548c499b202be50fc9d3df84dd4f/core/src/main/java/com/linecorp/armeria/server/VirtualHost.java#L492-L506

- We need to find a matching service to apply `CorsService` if it is statically wired.
- If `CorsService` is applied as a route decorator, `handlesCorsPreflight()` is false and `FallbackService` will be picked. However, `CorsService` will be applied by `InitialDispatcherService` regardless of the target service.

I will leave a comment about that in the code.

Reply:
Got it. Thanks!